jwql/instrument_monitors/pipeline_tools.py

```diff
@@ -0,0 +1,249 @@
+"""Various utility functions related to the JWST calibration pipeline
+
+Authors
+-------
+
+    - Bryan Hilbert
+
+Use
+---
+
+    This module can be imported as such:
+
+    >>> from jwql.instrument_monitors import pipeline_tools
+    pipeline_steps = pipeline_tools.completed_pipeline_steps(filename)
+ """
+
+from collections import OrderedDict
+import copy
+import numpy as np
+
+from astropy.io import fits
+from jwst import datamodels
+from jwst.dq_init import DQInitStep
+from jwst.dark_current import DarkCurrentStep
+from jwst.firstframe import FirstFrameStep
+from jwst.group_scale import GroupScaleStep
+from jwst.ipc import IPCStep
+from jwst.jump import JumpStep
+from jwst.lastframe import LastFrameStep
+from jwst.linearity import LinearityStep
+from jwst.persistence import PersistenceStep
+from jwst.ramp_fitting import RampFitStep
+from jwst.refpix import RefPixStep
+from jwst.rscd import RSCD_Step
+from jwst.saturation import SaturationStep
+from jwst.superbias import SuperBiasStep
+
+from jwql.utils.constants import JWST_INSTRUMENT_NAMES_UPPERCASE
+
+
+# Define the fits header keyword that accompanies each step
+PIPE_KEYWORDS = {'S_GRPSCL': 'group_scale', 'S_DQINIT': 'dq_init', 'S_SATURA': 'saturation',
+                 'S_IPC': 'ipc', 'S_REFPIX': 'refpix', 'S_SUPERB': 'superbias',
+                 'S_PERSIS': 'persistence', 'S_DARK': 'dark_current', 'S_LINEAR': 'linearity',
+                 'S_FRSTFR': 'firstframe', 'S_LASTFR': 'lastframe', 'S_RSCD': 'rscd',
+                 'S_JUMP': 'jump', 'S_RAMP': 'rate'}
+
+PIPELINE_STEP_MAPPING = {'dq_init': DQInitStep, 'dark_current': DarkCurrentStep,
+                         'firstframe': FirstFrameStep, 'group_scale': GroupScaleStep,
+                         'ipc': IPCStep, 'jump': JumpStep, 'lastframe': LastFrameStep,
+                         'linearity': LinearityStep, 'persistence': PersistenceStep,
+                         'rate': RampFitStep, 'refpix': RefPixStep, 'rscd': RSCD_Step,
+                         'saturation': SaturationStep, 'superbias': SuperBiasStep}
+
+# Readout patterns that have nframes != a power of 2. These readout patterns
+# require the group_scale pipeline step to be run.
+GROUPSCALE_READOUT_PATTERNS = ['NRSIRS2']
+
+
+def completed_pipeline_steps(filename):
+    """
```

Comment:
@bhilbert4 I think `sphinx` might freak out if line 61 and 62 here are not combined, e.g.:

```python
"""Return a list of completed pipeline steps for a given file.
```

Reply:
`sphinx` is so picky....